Assets/Scripts/Framework/Train/TrainEngine.cs

```diff
@@ -0,0 +1,218 @@
+
+        [HideInInspector]
+        [SyncVar(Channel = FishNet.Transporting.Channel.Reliable)]
+        private float _currentSpeed = 0f;
+        
+        private float _speedTypesCount = 0;
+
+        private Tween _speedTween;
+        private TrainController _trainController;
+
+        private void Awake() => _trainController = GetComponent<TrainController>();
+
+        private void Start()
+        {
+            _speedTypesCount = Enum.GetValues(typeof(TrainEngineSpeedTypes)).Length - 1;
+
+            // Initialize speed values dictionary.
+            _speedValues = new Dictionary<TrainEngineSpeedTypes, float>()
+            {
+                {TrainEngineSpeedTypes.high, _maxHighSpeed },
+                {TrainEngineSpeedTypes.low, _maxLowSpeed },
+                {TrainEngineSpeedTypes.medium, _maxMediumSpeed },
+                
+                {TrainEngineSpeedTypes.still, 0 },
+
+                {TrainEngineSpeedTypes.lowReverse, -_maxLowSpeed },
+                {TrainEngineSpeedTypes.mediumReverse, -_maxMediumSpeed },
+                {TrainEngineSpeedTypes.highReverse, -_maxHighSpeed },
+            };
+        }
+
+        /// <summary>
+        /// Helper method for checking if the train is moving forwards.
+        /// </summary>
+        /// <returns>True if the train is moving forward.</returns>
+        public bool IsTraveling()
+            => (int)CurrentEngineSpeedType > 3 || (int)CurrentTargetEngineSpeedType > 3;
+
+        /// <summary>
+        /// Helper method for checking if the train is moving backwards.
+        /// </summary>
+        /// <returns>True if the train moving backward.</returns>
+        public bool IsTravelingReverse()
+            => (int)CurrentEngineSpeedType < 3 || (int)CurrentTargetEngineSpeedType < 3;
+        
+        #region ServerRPCS
+        /// <summary>
+        /// Used to toggle direction of upcomming rail split.
+        /// </summary>
+        [ServerRpc(RequireOwnership = false)]
+        public void ToggleTrainDirection()
+            => OnTrainDirectionChanged(!CurrentSplitDirection);
+
+        /// <summary>
+        /// Used to set the train engine state.
+        /// </summary>
+        /// <param name="newState">New state of the train engine.</param>
+        [ServerRpc(RequireOwnership = false)]
+        public void SetTrainEngineState(TrainEngineState newState)
+            => OnTrainEngineStateChanged(newState);
+
+        /// <summary>
+        /// Method for increasing/decreasing train speed level.
+        /// </summary>
+        /// <param name="increment">Whether to increase or decrease speed.</param>
+        [ServerRpc(RequireOwnership = false)]
+        public void AdjustSpeed(bool increment)
+        {
+            TrainEngineSpeedTypes newTargetSpeed = (TrainEngineSpeedTypes)Mathf.Clamp
+                    ((int)CurrentTargetEngineSpeedType + (increment ? 1 : -1), 0, _speedTypesCount);
```

Comment:
I think this looks a bit nicer.
```suggestion
            TrainEngineSpeedTypes newTargetSpeed = (TrainEngineSpeedTypes)Mathf.Clamp(
                (int)CurrentTargetEngineSpeedType + (increment ? 1 : -1),
                0,
                _speedTypesCount
            );
```